pkg/dogstatsd/parse.go

```diff
@@ -0,0 +1,72 @@
+package dogstatsd
+
+import (
+	"bytes"
+	"strconv"
+	"strings"
+	"unsafe"
+)
+
+type messageType int
+
+const (
+	metricSampleType messageType = iota
+	serviceCheckType
+	eventType
+)
+
+var (
+	eventPrefix        = []byte("_e{")
+	serviceCheckPrefix = []byte("_sc")
+
+	fieldSeparator = []byte("|")
+	colonSeparator = []byte(":")
+	commaSeparator = []byte(",")
+
+	commaSeparatorString = ","
+)
+
+func findMessageType(message []byte) messageType {
+	if bytes.HasPrefix(message, eventPrefix) {
+		return eventType
+	} else if bytes.HasPrefix(message, serviceCheckPrefix) {
+		return serviceCheckType
+	}
+	// Note that random gibberish is interpreted as a metric since they don't
+	// contain any easily identifiable feature
+	return metricSampleType
+}
+
+// nextField returns the data found before the first fieldSeparator and
+// the remainder, as a no-heap alternative to bytes.Split.
+// If the separator is not found, the remainder is nil.
+func nextField(message []byte) ([]byte, []byte) {
+	sepIndex := bytes.Index(message, fieldSeparator)
+	if sepIndex == -1 {
+		return message, nil
+	}
+	return message[:sepIndex], message[sepIndex+1:]
+}
+
+func parseTags(rawTags []byte) []string {
+	if len(rawTags) == 0 {
+		return nil
+	}
+	return strings.Split(string(rawTags), commaSeparatorString)
+}
+
+// the std API does not have methods to do []byte => float parsing
+// we use this unsafe trick to avoid having to allocate one string for
+// every parsed float
+// see https://github.com/golang/go/issues/2632
+func parseFloat64(rawFloat []byte) (float64, error) {
+	return strconv.ParseFloat(*(*string)(unsafe.Pointer(&rawFloat)), 64)
+}
```

Comment:
I know it has been back ported and should work as expected, but since it's a "unsafe trick", I think it worth adding some unit test, just to test those 2 methods: `parseFloat64` and `parseInt64`